source/slang/slang-ast-builder.cpp

```diff
@@ -0,0 +1,196 @@
+}
+
+// !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!!! ASTBuilder !!!!!!!!!!!!!!!!!!!!!!!!!!!!!!
+
+ASTBuilder::ASTBuilder(SharedASTBuilder* sharedASTBuilder):
+    m_sharedASTBuilder(sharedASTBuilder)
+{
+    SLANG_ASSERT(sharedASTBuilder);
+}
+
+ASTBuilder::ASTBuilder():
+    m_sharedASTBuilder(nullptr)
+{
+}
+
+RefPtr<PtrType> ASTBuilder::getPtrType( RefPtr<Type>    valueType)
+{
+    return getPtrType(valueType, "PtrType").dynamicCast<PtrType>();
+}
+
+// Construct the type `Out<valueType>`
+RefPtr<OutType> ASTBuilder::getOutType(RefPtr<Type> valueType)
+{
+    return getPtrType(valueType, "OutType").dynamicCast<OutType>();
+}
+
+RefPtr<InOutType> ASTBuilder::getInOutType(RefPtr<Type> valueType)
+{
+    return getPtrType(valueType, "InOutType").dynamicCast<InOutType>();
+}
+
+RefPtr<RefType> ASTBuilder::getRefType(RefPtr<Type> valueType)
+{
+    return getPtrType(valueType, "RefType").dynamicCast<RefType>();
+}
+
+RefPtr<PtrTypeBase> ASTBuilder::getPtrType(RefPtr<Type> valueType, char const* ptrTypeName)
+{
+    auto genericDecl = m_sharedASTBuilder->findMagicDecl(ptrTypeName).dynamicCast<GenericDecl>();
+    return getPtrType(valueType, genericDecl);
+}
+
+RefPtr<PtrTypeBase> ASTBuilder::getPtrType(RefPtr<Type> valueType, GenericDecl* genericDecl)
+{
+    auto typeDecl = genericDecl->inner;
+
+    auto substitutions = create<GenericSubstitution>();
+    substitutions->genericDecl = genericDecl;
+    substitutions->args.add(valueType);
+
+    auto declRef = DeclRef<Decl>(typeDecl.Ptr(), substitutions);
+    auto rsType = DeclRefType::create(this, declRef);
+    return as<PtrTypeBase>(rsType);
+}
+
+RefPtr<ArrayExpressionType> ASTBuilder::getArrayType(Type* elementType, IntVal* elementCount)
+{
+    RefPtr<ArrayExpressionType> arrayType = create<ArrayExpressionType>();
+    arrayType->baseType = elementType;
+    arrayType->arrayLength = elementCount;
+    return arrayType;
+}
+
+RefPtr<VectorExpressionType> ASTBuilder::getVectorType(
+    RefPtr<Type>    elementType,
+    RefPtr<IntVal>  elementCount)
+{
+    auto vectorGenericDecl = m_sharedASTBuilder->findMagicDecl("Vector").as<GenericDecl>();
+        
+    auto vectorTypeDecl = vectorGenericDecl->inner;
+
+    auto substitutions = new GenericSubstitution();
+    substitutions->genericDecl = vectorGenericDecl.Ptr();
+    substitutions->args.add(elementType);
+    substitutions->args.add(elementCount);
+
+    auto declRef = DeclRef<Decl>(vectorTypeDecl.Ptr(), substitutions);
+
+    return DeclRefType::create(this, declRef).as<VectorExpressionType>();
+}
+
+RefPtr<TypeType> ASTBuilder::getTypeType(Type* type)
+{
+    return create<TypeType>(type);
+}
+
+SyntaxNodeBase* ASTBuilder::createInstanceOfSyntaxClassByName(const String& name)
+{
+    if (0) {}
+#define CASE(NAME) \
+        else if(name == #NAME) return create<NAME>()
+
+    CASE(GLSLBufferModifier);
+    CASE(GLSLWriteOnlyModifier);
+    CASE(GLSLReadOnlyModifier);
+    CASE(GLSLPatchModifier);
+    CASE(SimpleModifier);
```

Comment:
Are these really the only cases that need create-from-name functionality?

Reply:
I guess that's kind of silly - as we have the reflection mechansism that creates any of them. I just copied it over...

Reply:
Ok removed that.